torch/utils/pytree.py

```diff
@@ -0,0 +1,767 @@
+"""
+Contains utility functions for working with nested python data structures.
+
+A *pytree* is Python nested data structure. It is a tree in the sense that
+nodes are Python collections (e.g., list, tuple, dict) and the leaves are
+Python values. Furthermore, a pytree should not contain reference cycles.
+
+pytrees are useful for working with nested collections of Tensors. For example,
+one can use `tree_map` to map a function over all Tensors inside some nested
+collection of Tensors and `tree_leaves` to get a flat list of all Tensors
+inside some nested collection. pytrees are helpful for implementing nested
+collection support for PyTorch APIs.
+"""
+
+import functools
+import pickle
+from typing import (
+    Any,
+    Callable,
+    Iterable,
+    List,
+    Optional,
+    overload,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+)
+
+import optree
```

Comment:
Should we be guarding this import with a try/catch and fallback to the python pytree if the dependency doesn't exist?

Reply:
`torch.utils.pytree` is not imported when we write `import torch`. Maybe we can guard this like we do for `numpy` and prompt a warning.

Reply:
The issue is the next steps. i imagine the next steps are something like "incrementally replace all uses of `torch.utils._pytree` with `torch.utils.pytree`", right?

Right now this isn't a problem, but once a file actually starts using `torch.utils.pytree`, this is an issue. I think we need to guard on `optree` being present and if it isn't, fall back to using `torch.utils._pytree`.

Reply:
I'll submit a follow-up PR to resolve this.

Reply:
Thanks @XuehaiPan! I'm excited to start porting PyTorch to use optree :) It's been a long time delayed.